priv/repo/migrations/20180426164341_create_hourly_downloads.exs

```diff
@@ -0,0 +1,56 @@
+defmodule Castle.Repo.Migrations.CreateHourlyDownloads do
+  use Ecto.Migration
+
+  def change do
+    create table(:hourly_downloads, primary_key: false) do
+      add :podcast_id, :integer, null: false
+      add :episode_id, :uuid, null: false
+      add :dtim, :utc_datetime, null: false
+      add :count, :integer, null: false
+    end
+
+    execute """
+    CREATE OR REPLACE FUNCTION create_hourly_downloads_partition() RETURNS trigger AS
+    $$
+      DECLARE
+        partition_start DATE;
+        partition_end DATE;
+        partition TEXT;
+      BEGIN
+        partition_start := DATE_TRUNC('MONTH', NEW.dtim);
+        partition_end := partition_start + INTERVAL '1 MONTH';
+        partition := 'hourly_downloads_' || to_char(NEW.dtim,'YYYYMM');
+        IF NOT EXISTS(SELECT relname FROM pg_class WHERE relname=partition) THEN
+          RAISE NOTICE 'A partition has been created %',partition;
+          EXECUTE 'CREATE TABLE ' || partition || ' (CHECK (dtim >= DATE ''' || partition_start || ''' AND dtim < DATE ''' || partition_end || '''), CONSTRAINT ' || partition || '_uniq UNIQUE (episode_id, dtim)) INHERITS (hourly_downloads);';
+          EXECUTE 'CREATE INDEX ' || partition || '_podcast_id_index ON ' || partition || ' (podcast_id);';
+          EXECUTE 'CREATE INDEX ' || partition || '_episode_id_index ON ' || partition || ' (episode_id);';
+          EXECUTE 'CREATE INDEX ' || partition || '_dtim_index ON ' || partition || ' (dtim);';
```

Comment:
Not entirely sure what all indices we'll need on the partition tables.  But they have to be added here by this trigger - they're not inherited from the parent table.

Any thoughts on what we'll need for our queries?  Guessing we'll most ofter be looking at a specific podcast/episode, and filtering by a date range.

Reply:
I think your query instincts are correct, at least as far as I have imagined. I will read up on on postgres query optimization + partitioning. Since it's partitioned by day and queried by podcast or episode + date range, I may have been inclined to include the dtim in the podcast and episode indexes for partitions, but I can't give solid advice on that just yet.

Reply:
If you've got this running locally, and querying from BQ-prod ... you can rollup a couple of days, and then `explain` a few queries against it.

```
mix castle.rollup.downloads -d 20180502
mix castle.rollup.downloads -d 20180403
mix castle.rollup.downloads -d 20180304
mix castle.rollup.downloads -d 20180302
mix castle.rollup.downloads -d 20180201
```

I was imagining queries like:

```
SELECT dtim, SUM(count) FROM hourly_downloads WHERE dtim >= '2018-04-01' AND dtim < '2018-04-15' AND podcast_id = 70 GROUP BY dtim ORDER BY dtim asc;